### PR TITLE
Add baseURL to ORK[1]WebViewStep to allow getUserMedia to work

### PR DESCRIPTION
**Pull request by eschramm**
Fixes https://github.com/CareEvolution/RKStudio-Participant/issues/924. `WKWebView` now supports `getUserMedia()` which allows access to the camera. I found that if a `WKWebView` is fed a `URLRequest` to a URL containing the exact same code as our WebViewStep, it works just fine. Apparently, when using `loadHTMLString:`, the `getUserMedia()` call will work so long as any arbitrary URL is set as `baseURL` in that call - i.e. it cannot be `nil`. To prevent potential regression issues, this solution simply exposes `baseURL` as a nullable property on `ORK(1)WebViewStep`

## Testing:

Attempting the following survey should not properly show a live photo preview (possibly prompting for permission to use camera if not yet granted). Will require corresponding PRs in [CEVResearchKit](https://github.com/CareEvolution/CEVResearchKit/pull/242) and [RKStudio-Participant](https://github.com/CareEvolution/RKStudio-Participant/pull/926).

```json
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "steps": [
    {
      "type": "InstructionStep",
      "text": "This is the image capture step",
      "title": "",
      "identifier": "New Step 1"
    },
    {
      "type": "WebViewStep",
      "text": "",
      "title": "",
      "identifier": "New Step 4",
      "html": "<iframe allowfullscreen\n                style=\"border: none; padding: 0; margin: 0; width: 100%; display: block;\"\n                sandbox=\"allow-scripts allow-same-origin allow-presentation allow-popups\"\n                allow=\"camera\"\n                src=\"\" tabindex=\"0\"></iframe>\n\n<html>\n<head>\n\n  <link rel=\"stylesheet\" href=\"https://rkstudio-customer-assets.s3.amazonaws.com/Shared/font-awesome/4.7.0/css/font-awesome.min.css\" />\n\n  <style type=\"text/css\">\n    body {\n      margin: 0;\n      padding: 10px;\n    }\n\n    @media screen and (max-width: 600px) {\n      body {\n        padding: 5px;\n      }\n    }\n\n    input {\n      border: none;\n      padding: 10px;\n      margin-top: 10px;\n      width: 100%;\n      box-sizing: border-box;\n      font-size: 16px;\n      line-height: 20px;\n      border-top: solid 1px #ddd;\n      border-bottom: solid 1px #ddd;\n    }\n\n    input:focus {\n      outline: none;\n    }\n\n    button {\n      width: 100%;\n      box-sizing: border-box;\n      padding: 16px;\n      background: #5381c3;\n      font-size: 17px;\n      color: #FFF;\n      border: none;\n      border-radius: 5px;\n      cursor: pointer;\n      font-weight: bold;\n      margin-top: 20px;\n    }\n\n    button:disabled {\n      background: #FFF !important;\n      color: #999 !important;\n      border-color: #ccc !important;\n      cursor: auto;\n    }\n\n    div {\n      text-align: center;\n    }\n\n    #video {\n      border: 1px solid black;\n      box-shadow: 2px 2px 3px black;\n    }\n\n    #photo {\n      border: 1px solid black;\n      box-shadow: 2px 2px 3px black;\n      display: none;\n    }\n\n    #canvas {\n      display: none;\n    }\n\n    .camera {\n      position: relative;\n    }\n\n    .output {\n      text-align: center;\n    }\n\n    .strip-placeholder {\n      border: 2px solid #999;\n      width: 80%;\n      left: 10%;\n      top: 20%;\n      height: 20px;\n      position: absolute;\n    }\n\n    .card-frame {\n      border: 3px dotted #999;\n      width: 94%;\n      height: 94%;\n      position: absolute;\n      top: 2%;\n      left: 2.3%;\n    }\n\n    .barcode-placeholder {\n      position: absolute;\n    }\n\n    .camera.vertical .strip-placeholder {\n      width: 5%;\n      height: 88.6%;\n      top: 5.7%;\n      left: 22.7%;\n    }\n\n    .camera.vertical .card-frame {\n      width: 92%;\n      top: 2.7%;\n      left: 3.2%;\n    }\n\n    .camera.vertical .barcode-placeholder {\n      left: 54.4%;\n      width: 19%;\n      height: 31.4%;\n      top: 34.3%;\n    }\n\n    #processing-loading,\n    #analyzing-loading {\n      display: none;\n    }\n  </style>\n  <script src=\"https://rkstudio-customer-assets.s3.amazonaws.com/Shared/jquery/jquery-3.4.1.min.js\"></script>\n  <script src=\"https://rkstudio-customer-assets.s3.amazonaws.com/Shared/lodash/lodash-4.17.15.min.js\"></script>\n</head>\n<body>\n  \n  <div>Test Result</div>\n  <div class=\"camera\">\n    <video id=\"video\" playsinline=\"true\">Video stream not available.</video>\n    <div class=\"card-frame\">\n      <div class=\"strip-placeholder\"></div>\n      <div class=\"barcode-placeholder\">\n        <?xml version=\"1.0\" encoding=\"utf-8\"?>\n        <svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" style=\"width: 100%; height: 100%;\">\n          <g transform=\"translate(0,240) scale(0.5 0.5) rotate(-90)\">\n            <polygon points=\"45 0 45 240 49 240 49 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"49 0 49 240 54 240 54 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"58 0 58 240 63 240 63 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"72 0 72 240 77 240 77 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"77 0 77 240 81 240 81 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"81 0 81 240 86 240 86 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"95 0 95 240 99 240 99 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"113 0 113 240 117 240 117 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"122 0 122 240 127 240 127 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"127 0 127 240 131 240 131 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"145 0 145 240 149 240 149 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"158 0 158 240 163 240 163 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"181 0 181 240 185 240 185 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"185 0 185 240 190 240 190 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"195 0 195 240 199 240 199 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"199 0 199 240 204 240 204 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"204 0 204 240 208 240 208 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"213 0 213 240 217 240 217 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"222 0 222 240 226 240 226 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"226 0 226 240 231 240 231 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"244 0 244 240 249 240 249 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"258 0 258 240 263 240 263 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"267 0 267 240 272 240 272 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"272 0 272 240 276 240 276 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"294 0 294 240 299 240 299 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"303 0 303 240 308 240 308 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"308 0 308 240 313 240 313 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"313 0 313 240 317 240 317 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"331 0 331 240 335 240 335 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"335 0 335 240 340 240 340 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"344 0 344 240 349 240 349 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"353 0 353 240 358 240 358 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"358 0 358 240 362 240 362 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"362 0 362 240 367 240 367 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"371 0 371 240 376 240 376 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"376 0 376 240 381 240 381 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"381 0 381 240 385 240 385 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"385 0 385 240 390 240 390 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"394 0 394 240 399 240 399 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"399 0 399 240 403 240 403 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"417 0 417 240 421 240 421 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"421 0 421 240 426 240 426 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"426 0 426 240 430 240 430 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"435 0 435 240 440 240 440 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"444 0 444 240 449 240 449 0\" fill=\"rgb(153,153,153)\" />\n            <polygon points=\"449 0 449 240 453 240 453 0\" fill=\"rgb(153,153,153)\" />\n          </g>\n        </svg>\n      </div>\n    </div>\n  </div>\n  <button id=\"capture-image\" onclick=\"captureImage()\">Capture Image</button>\n  <canvas id=\"canvas\"></canvas>\n  <div class=\"output\">\n    <img id=\"photo\" alt=\"The screen capture will appear in this box.\">\n  </div>\n  <br />\n  <div id=\"processing-loading\">\n    <i class=\"fa fa-refresh fa-spin\" style=\"color:#999\"></i>\n    Processing Image...\n  </div>\n  <div style=\"text-align: center;\"><img src=\"\" id=\"processed-image\" /></div>\n  <br />\n  <div id=\"results\"></div>\n  <div id=\"analyzing-loading\">\n    <i class=\"fa fa-refresh fa-spin\" style=\"color:#999\"></i>\n    Analyzing Results...\n  </div>\n  <div id=\"log\"></div>\n  <script>\n\n    var streaming = false;\n\n    var video = null;\n    var canvas = null;\n    var photo = null;\n    var camera = null;\n\n    function startup() {\n      video = document.getElementById('video');\n      canvas = document.getElementById('canvas');\n      photo = document.getElementById('photo');\n      camera = document.getElementsByClassName('camera')[0];\n\n      var windowWidth = document.documentElement.clientWidth;\n      var windowHeight = document.documentElement.clientHeight;\n\n      var isVertical = windowHeight > windowWidth;\n\n      var aspectRatio = isVertical ? 4 / 3 : 3 / 4;\n\n      if (isVertical) {\n        camera.classList.add(\"vertical\");\n      }\n\n      console.log(camera.offsetWidth);\n      console.log($(camera).width());\n\n      navigator.mediaDevices.getUserMedia({\n        video: {\n          width: {\n            ideal: 5000\n          },\n          aspectRatio: aspectRatio,\n          facingMode: { ideal: \"environment\" }\n        },\n        audio: false\n      })\n        .then(function(stream) {\n          video.srcObject = stream;\n          video.play();\n          $(video).attr(\"width\", camera.offsetWidth - 8);\n          $(photo).attr(\"width\", camera.offsetWidth - 8);\n        });\n\n      video.addEventListener('canplay', function(ev) {\n        if (!streaming) {\n          canvas.setAttribute('width', video.videoWidth);\n          canvas.setAttribute('height', video.videoHeight);\n          streaming = true;\n        }\n      }, false);\n\n      clearphoto();\n    }\n\n    function clearphoto() {\n      var context = canvas.getContext('2d');\n      context.fillStyle = \"#AAA\";\n      context.fillRect(0, 0, canvas.width, canvas.height);\n\n      var data = canvas.toDataURL('image/png');\n      photo.setAttribute('src', data);\n    }\n\n    function captureImage() {\n      var context = canvas.getContext('2d');\n      if (streaming) {\n        canvas.width = video.videoWidth;\n        canvas.height = video.videoHeight;\n        context.drawImage(video, 0, 0, video.videoWidth, video.videoHeight);\n\n        console.log('Taking photo with width: ' + video.videoWidth + ' and height ' + video.videoHeight);\n\n        var data = canvas.toDataURL('image/png');\n        photo.setAttribute('src', data);\n\n        $(\".camera\").hide();\n        $(\"#capture-image\").hide();\n        photo.style.display = \"block\";\n\n        canvas.toBlob(function(blob) {\n\n          var reader = new FileReader();\n          reader.onload = function(e) {\n            console.log(e.target.result.byteLength);\n            processImage(e.target.result);\n          };\n          reader.readAsArrayBuffer(blob);\n\n        }, \"image/jpeg\");\n      } else {\n        clearphoto();\n      }\n    }\n\n    startup();\n  </script>\n</body>\n</html>",
      "height": 150,
      "baseURL": "https://mydatahelps.org"
    },
    {
      "type": "CompletionStep",
      "text": "Done",
      "title": "",
      "identifier": "New Step 3"
    }
  ],
  "styles": {
    "progressBarColor": "#5381c3",
    "nextButtonBackgroundColor": "#5381c3",
    "nextButtonTextColor": "#FFFFFF"
  }
}
```

### CODE REVIEW

File: ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.h [hunk 1]
```diff
@@ -31,15 +31,15 @@
 @import UIKit;
 @import WebKit;
 #import <ORK1Kit/ORK1Defines.h>
-#import "ORK1StepViewController.h"
+#import <ORK1Kit/ORK1StepViewController.h>
```
Comment (overcyn) [inline ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.h:34]: You changed it here but not in ORKWebViewStepViewController. Does it matter?

File: ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.m [hunk 2]
```diff
@@ -112,12 +112,12 @@ - (instancetype)initWithStep:(ORK1Step *)step {
...
         };
         
-        _webView = [[ORK1WebViewPreloader shared] webViewForKey:step.identifier];
+        _webView = [[ORK1WebViewPreloader shared] webViewForKey:step.identifier baseURL:[self webViewStep].baseURL];
         [_webView.configuration.userContentController addScriptMessageHandler:_scriptMessageHandlerImpl name:@"ResearchKit"];
         _webView.frame = self.view.bounds;
         _webView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
         _webView.navigationDelegate = self;
-        [_webView loadHTMLString:[self webViewStep].html baseURL:nil];
+        [_webView loadHTMLString:[self webViewStep].html baseURL:[self webViewStep].baseURL];
```
Comment (overcyn) [inline ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.m:120]: In ORK1WebViewStepViewController.stepDidChange and ORKWebViewStepViewController.stepDidChange it calls `[_webView loadHTMLString:[self webViewStep].html baseURL:nil];` without a baseURL. Does it need it there?

**Review by overcyn — COMMENTED** [review]
What are the advantages for having a survey configurable baseURL vs something like `participantContext.site.uri`?

**Comment by chrisnowak** [comment]
Awesome that you were able to figure this out!

Wondering though what kinds of regression issues we are expecting?  is there concern there might be relative urls or something that might resolve?  Seems like just hardcoding it to some value (e.g. https://careevolutionframe.com/ ) would be easiest rather than having to adjust the baseUrl of the survey behind the scenes

File: ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.h (hunk 1 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.h:34]: That was to silence a warning - strangely it only appeared in the ORK1Kit version.

File: ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.m (hunk 2 above)
Comment (eschramm) [reply ORK1Kit/ORK1Kit/Common/ORK1WebViewStepViewController.m:120]: Good catch, thank you!

**Comment by eschramm** [comment]
Yeah I debated defaulting in an arbitrary URL but I wasn't sure what the effect COULD be given we probably have hundreds of WebViewSteps live. I figured to ship it sooner, let's make it still default to nil, but configure it to something while we work on understanding what the effects may be. As usual, there's not much [documentation](https://developer.apple.com/documentation/webkit/wkwebview/1415004-loadhtmlstring?language=occ) - I would assume it should only affect any simple links in the static html. If we think it'd be best to default it as `participantContext.site.uri` or https://careevolutionframe.com, I'm game to set the default state to that vs `nil`.

**Comment by chrisnowak** [comment]
Hmm yeah, I guess let's just keep it as-is (setting base url as a property).  Then in the future we can default it to something when we have confidence it won't break stuff

**Comment by overcyn** [comment]
There is still the `[_webView loadHTMLString:[self webViewStep].html baseURL:nil];` in `[ORK1WebViewStepViewController stepDidChange]`. Otherwise LGTM.

**Pull request merged**